File: Day06/Python/Day06.py
import os
import time
import numpy as np

# Read and parse the input
f = open("../input.txt", 'r')
file = f.read().strip()
grid = np.array([[j for j in i] for i in file.split("\n")])
rows, cols = len(grid), len(grid[0])
startingPoint = np.where(grid == '^')

x, y = startingPoint[0][0], startingPoint[1][0]

# Define movement directions
rotations = ([-1, 0],   # Moving UP     ↑
            [0, 1],     # Moving RIGHT  →
            [1, 0],     # Moving DOWN   ↓
            [0, -1])    # Moving LEFT   ←
rotationID = 0
loops = 0  # Accumulator for Part Two

def prettyPrint(grid):
    """Display the grid with 'X' highlighted."""
    RESET = '\033[0m'
    BLUE = '\033[34m'
    colored_result = '\n'.join(
        ''.join(BLUE + 'X' + RESET if char == 'X' else char for char in row)
        for row in grid)
    time.sleep(.1)
    os.system('clear')
    print(colored_result)

def traverse(grid, x, y, obsX, obsY, rotationID):
    """Traverse the grid and detect loops, incrementing the global loops counter."""
    startX, startY, startRot = x, y, rotationID
    seen = set()  # Keep track of visited (x, y, rotationID)
    TTL = 130**2  # Arbitrary limit to prevent infinite loops
    
    # Mark obstacle
    grid[obsX, obsY] = "#"
    while TTL > 0:
        TTL -= 1
        state = (x, y, rotationID)
        if state in seen:
            return True  # Loop detected
        seen.add(state)
        
        nextX, nextY = x + rotations[rotationID][0], y + rotations[rotationID][1]
        if nextX < 0 or nextX >= rows or nextY < 0 or nextY >= cols:
            return False  # Out of bounds
        
        if grid[nextX, nextY] == "#":
            rotationID = (rotationID + 1) % 4
        else:
            x, y = nextX, nextY

    return False  # Exceeded time-to-live

# Part One: Guard Route
obstacles = []
visited_positions = set()
while True:
    grid[x, y] = 'X'
    visited_positions.add((x, y))
    nextX, nextY = x + rotations[rotationID][0], y + rotations[rotationID][1]
    
    if (nextX < 0 or nextX >= rows) or (nextY < 0 or nextY >= cols):
        break  # Out of bounds
    
    if grid[nextX, nextY] == "#":
        rotationID = (rotationID + 1) % 4
        obstacles.append((int(nextX), int(nextY)))
        x, y = x + rotations[rotationID][0], y + rotations[rotationID][1]
    else:
        x, y = nextX, nextY

# Part Two: Finding Valid Obstruction Positions
valid_positions = 0
for obsX, obsY in visited_positions:
    if (obsX, obsY) == (startingPoint[0][0], startingPoint[1][0]) or grid[obsX, obsY] == "#":
        continue  # Skip guard's starting position or existing walls
    temp_grid = grid.copy()
    if traverse(temp_grid, startingPoint[0][0], startingPoint[1][0], obsX, obsY, 0):
        valid_positions += 1

# Final results
print("#" * 50)
print(f"Advent of Code 2024\n\t- Day 06\n\t\t★  Result: {np.count_nonzero(grid == 'X')}")
print(f"\t\t★★ Result: {valid_positions}")
print("#" * 50)

File: Day06/C++/Day06.cpp
#include <iostream>
#include <vector>
#include <set>
#include <fstream>
#include <string>
#include <tuple>
#include <chrono>
#include <thread>
#include <unordered_set>
#include <omp.h>   // OpenMP header for parallelization
#include <atomic>  // For thread-safe counting

using namespace std;

// Direction vectors for movements: UP, RIGHT, DOWN, LEFT
const vector<pair<int, int>> rotations = {
    {-1, 0},  // UP
    {0, 1},   // RIGHT
    {1, 0},   // DOWN
    {0, -1}   // LEFT
};

// Function to print grid (removed colored output for simplicity)
void prettyPrint(const vector<vector<char>> &grid) {
    for (const auto &row : grid) {
        for (char c : row) {
            cout << c;
        }
        cout << endl;
    }
}

// Parallelized version of traverse function
bool traverse(const vector<vector<char>> &grid, int startX, int startY, int obsX, int obsY, int rotationID) {
    int rows = grid.size();
    int cols = grid[0].size();
    
    // Create a copy of the grid to modify
    vector<vector<char>> workGrid = grid;
    workGrid[obsX][obsY] = '#';
    
    set<tuple<int, int, int>> seen;
    int x = startX, y = startY;

    while (true) {
        auto state = make_tuple(x, y, rotationID);
        if (seen.find(state) != seen.end()) {
            return true; // Loop detected
        }
        seen.insert(state);

        int nextX = x + rotations[rotationID].first;
        int nextY = y + rotations[rotationID].second;

        if (nextX < 0 || nextX >= rows || nextY < 0 || nextY >= cols) {
            return false; // Out of bounds
        }

        if (workGrid[nextX][nextY] == '#') {
            rotationID = (rotationID + 1) % 4; // Rotate clockwise
        } else {
            x = nextX;
            y = nextY;
        }
    }
}

int main() {
    auto start_time = chrono::high_resolution_clock::now();

    // Read input from file
    ifstream infile("../input.txt");
    vector<string> file_lines;
    string line;

    while (getline(infile, line)) {
        file_lines.push_back(line);
    }

    int rows = file_lines.size();
    int cols = file_lines[0].size();
    vector<vector<char>> grid(rows, vector<char>(cols));
    int startX = 0, startY = 0;

    // Populate grid and find starting position
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            grid[i][j] = file_lines[i][j];
            if (grid[i][j] == '^') {
                startX = i;
                startY = j;
            }
        }
    }

    // Part One: Simulate guard route (same as original)
    int x = startX, y = startY, rotationID = 0;
    vector<pair<int, int>> obstacles;
    set<pair<int, int>> visited_positions;

    while (true) {
        grid[x][y] = 'X';
        visited_positions.insert({x, y});
        int nextX = x + rotations[rotationID].first;
        int nextY = y + rotations[rotationID].second;

        if (nextX < 0 || nextX >= rows || nextY < 0 || nextY >= cols) {
            break; // Out of bounds
        }

        if (grid[nextX][nextY] == '#') {
            rotationID = (rotationID + 1) % 4;
            obstacles.emplace_back(nextX, nextY);
            x = x + rotations[rotationID].first;
            y = y + rotations[rotationID].second;
        } else {
            x = nextX;
            y = nextY;
        }
    }

    // Part Two: Parallelize finding valid obstruction positions
    std::atomic<int> valid_positions(0);

    // Use OpenMP to parallelize the loop
    #pragma omp parallel
    {
        // Create thread-local grid to avoid race conditions
        vector<vector<char>> thread_grid = grid;

        // Distribute iterations across threads
        #pragma omp for
        for (int idx = 0; idx < visited_positions.size(); ++idx) {
            auto it = std::next(visited_positions.begin(), idx);
            int obsX = it->first;
            int obsY = it->second;

            // Skip starting position or existing walls
            if ((obsX == startX && obsY == startY) || thread_grid[obsX][obsY] == '#') {
                continue;
            }

            // Check if this position allows a valid traversal
            if (traverse(thread_grid, startX, startY, obsX, obsY, 0)) {
                valid_positions++;
            }
        }
    }

    // Final Results
    int visited_count = visited_positions.size();

    cout << string(50, '#') << endl;
    cout << "Advent of Code 2024\n\t- Day 06" << endl;
    cout << "\t\t★  Result: " << visited_count << endl;
    cout << "\t\t★★ Result: " << valid_positions.load() << endl;
    cout << string(50, '#') << endl;

    auto end_time = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end_time - start_time;
    cout << "Execution Time: " << elapsed.count() << " seconds" << endl;

    return 0;
}